Fold weight permutation inside quantized conv operator

dzhulgakov opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26339 Serialization for per channel qtensor
* #26243 Fix _empty_per_channel_affine_quantized to be less hacky
* #26242 Fold activation permutation inside quantized conv operator
* **#26241 Fold weight permutation inside quantized conv operator**
* #26240 Implement more support for per-channel quantization

According to https://github.com/pytorch/pytorch/issues/19092 we always keep NCHW order and do handling inside the kernels. This PR fixes it for weights of the qconv by using MemoryLayout mechanism.

Differential Revision: [D17443219](https://our.internmc.facebook.com/intern/diff/D17443219)

Commit:
permute for weights

Commit:
Update on "Fold weight permutation inside quantized conv operator"


According to https://github.com/pytorch/pytorch/issues/19092 we always keep NCHW order and do handling inside the kernels. This PR fixes it for weights of the qconv by using MemoryLayout mechanism.

[ghstack-poisoned]

jerryzh168 reviewed (approved):
LGTM

jerryzh168 commented:
What about quantize linear ops?

dzhulgakov commented:
What do you mean about linear? Linear ops don't have memory order

Also cc @VitalyFedyunin to verify that memory format stuff looks good

supriyar commented on test/test_quantized_nn_mods.py:
```diff
@@ -65,7 +65,7 @@ def test_conv_api(self, use_bias):
 
         b = torch.randn(oC, dtype=torch.float32) if use_bias else None
         q_bias = torch.quantize_linear(b, scale=1.0 / 1024, zero_point=0, dtype=torch.qint32) if use_bias else None
-        q_filters_ref = torch.ops.quantized.conv_prepack(qw.permute([0, 2, 3, 1]),
```
Don't you also need similar changes to the tests in test_quantized.py?

dzhulgakov replied:
oh, I think I put it accidentally in https://github.com/pytorch/pytorch/pull/26242 - let me just squash them together

Commit:
Update on "Fold weight permutation inside quantized conv operator"


According to https://github.com/pytorch/pytorch/issues/19092 we always keep NCHW order and do handling inside the kernels. This PR fixes it for weights of the qconv by using MemoryLayout mechanism.

[ghstack-poisoned]

Commit:
Update on "Fold weight permutation inside quantized conv operator"


According to https://github.com/pytorch/pytorch/issues/19092 we always keep NCHW order and do handling inside the kernels. This PR fixes it for weights of the qconv by using MemoryLayout mechanism.

[ghstack-poisoned]

VitalyFedyunin commented:
Please don't land, I need to read it carefully.

VitalyFedyunin reviewed (approved):
Memory Format manipulations looks good!

Commit:
Update on "Fold weight permutation inside quantized conv operator"


According to https://github.com/pytorch/pytorch/issues/19092 we always keep NCHW order and do handling inside the kernels. This PR fixes it for weights of the qconv by using MemoryLayout mechanism.

Differential Revision: [D17443219](https://our.internmc.facebook.com/intern/diff/D17443219)

[ghstack-poisoned]

Commit:
Update on "Fold weight permutation inside quantized conv operator"


According to https://github.com/pytorch/pytorch/issues/19092 we always keep NCHW order and do handling inside the kernels. This PR fixes it for weights of the qconv by using MemoryLayout mechanism.

Differential Revision: [D17443219](https://our.internmc.facebook.com/intern/diff/D17443219)

[ghstack-poisoned]

Commit:
Update on "Fold weight permutation inside quantized conv operator"


According to https://github.com/pytorch/pytorch/issues/19092 we always keep NCHW order and do handling inside the kernels. This PR fixes it for weights of the qconv by using MemoryLayout mechanism.

Differential Revision: [D17443219](https://our.internmc.facebook.com/intern/diff/D17443219)

[ghstack-poisoned]

supriyar commented on test/test_quantized.py:
```diff
@@ -1177,18 +1175,15 @@ def test_qconv_unpack(self, X, strideH, strideW, padH, padW, channelwise):
-        W_KRSC = W.permute([0, 2, 3, 1]).contiguous()
         if channelwise:
-            W_q = torch.quantize_linear_per_channel(W_KRSC,
+            W_q = torch.quantize_linear_per_channel(W,
```
Thanks for changing this too. Did you test it locally by setting `PYTORCH_TEST_WITH_QNNPACK` ?

dzhulgakov replied:
Yes, I did :)

supriyar commented on aten/src/ATen/native/quantized/cpu/qconv_unpack.cpp:
```diff
@@ -55,9 +57,10 @@ class QConvUnpackWeightsInt8 final : public c10::OperatorKernel {
           scales.toType(kDouble),
           zero_points.toType(kLong),
-          {output_channels, kernel_h, kernel_w, C_per_G},
+          {output_channels, C_per_G, kernel_h, kernel_w},
```
Do we need to change it similarly for QNNPACK as well below since we store it as `MemoryFormat::ChannelsLast` in the packed struct?

dzhulgakov replied:
No, you store the original Tensor and it's fine to return Tensor in ChannelLast format (it'd just happen to be non-contiguous but still semantically correct)

Commit:
Update on "Fold weight permutation inside quantized conv operator"


According to https://github.com/pytorch/pytorch/issues/19092 we always keep NCHW order and do handling inside the kernels. This PR fixes it for weights of the qconv by using MemoryLayout mechanism.

Differential Revision: [D17443219](https://our.internmc.facebook.com/intern/diff/D17443219)

[ghstack-poisoned]
